src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/UpToDateCheckImplicitConfiguredInput.cs

```diff
@@ -411,7 +445,11 @@ void AddKeys(ImmutableDictionary<string, ImmutableDictionary<string, ImmutableAr
 
                 var itemsByKindBySet = new Dictionary<string, Dictionary<string, HashSet<string>>>(BuildUpToDateCheck.SetNameComparer);
 
-                foreach ((string item, IImmutableDictionary<string, string> metadata) in projectChangeDescription.After.Items)
+                var after = projectChangeDescription.After.Items as IDataWithOriginalSource<KeyValuePair<string, IImmutableDictionary<string, string>>>;
+
+                Assumes.NotNull(after);
```

Comment:
This is the first approach for ensuring stability: The use of `IDataWithOriginalSource<T>` on the snapshot, to preserve order.